### PR TITLE
[Feature] Support DiffusionDet Inference in Projects

### PR DESCRIPTION
**Pull request by BIGWangYuDong**
Ready for the inference-time performance of DiffusionDet in MMDetection projects matches that in the original paper(repo)

**Note:**
During inference time, DiffusionDet will randomly create boxes, which may affect the results (AP).

If hard-setting seeds=0 before generating random boxes, the AP will be same:
``` python
# hard set seed=0 before generating random boxes
seed = 0
random.seed(seed)
# np.random.seed(seed)
torch.manual_seed(seed)
# torch.cuda.manual_seed(seed)
torch.cuda.manual_seed_all(seed)

...

noise_bboxes_raw = torch.randn(
    (self.num_proposals, 4),
    device=device)

...
```

| repo  | mAP |
|:------:|:-----:|
|  DifussionDet  (without any changing)  |  45.5 |
| DifussionDet (hard set seed=0) | 45.66 |
| mmdet repo  (hard set seed=0)  | 45.7 |
| mmdet repo  (random seed)  | 45.6~45.8 |

In addition, we give a script to convert DiffusionDet-resnet50 to MMDet version. 

``` shell
python projects/DiffusionDet/model_converters/diffusiondet_resnet_to_mmdet.py ${PATH_DOWNLOAD_FROM_DIFFUSIONDET} ${CONVERTED_SAVING_PATH}
```

### CODE REVIEW

File: configs/_base_/datasets/coco_detection.py [hunk 1]
```diff
@@ -10,16 +10,24 @@
 #     }))
```
Comment (ZwwWayne) [inline configs/_base_/datasets/coco_detection.py:10]: configs in _base_ should not be modified.

File: mmdet/datasets/base_det_dataset.py [hunk 2]
```diff
@@ -111,3 +111,49 @@ def get_cat_ids(self, idx: int) -> List[int]:
         """
         instances = self.get_data_info(idx)['instances']
         return [instance['bbox_label'] for instance in instances]
+
+    def __getitem__(self, idx: int) -> dict:
+        """Get the idx-th image and data information of dataset after
+        ``self.pipeline``, and ``full_init`` will be called if the dataset has
```
Comment (ZwwWayne) [inline mmdet/datasets/base_det_dataset.py:117]: What is modified? We should add description

File: projects/DiffusionDet/README.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+```BibTeX
+@article{chen2022diffusiondet,
+      title={DiffusionDet: Diffusion Model for Object Detection},
+      author={Chen, Shoufa and Sun, Peize and Song, Yibing and Luo, Ping},
+      journal={arXiv preprint arXiv:2211.09788},
+      year={2022}
+}
+```
+
+## Checklist
```
Comment (hhaAndroid) [inline projects/DiffusionDet/README.md:62]: Can the following be deleted?

File: projects/DiffusionDet/configs/diffusiondet_r50_fpn_500-proposals_crop-ms-480-800-450k_coco.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        score_thr=0.5,
+        min_bbox_size=0,
+        nms=dict(type='nms', iou_threshold=0.5),
+    ))
+
+backend = 'pillow'
+train_pipeline = [
+    dict(
+        type='LoadImageFromFile',
+        file_client_args={{_base_.file_client_args}},
```
Comment (hhaAndroid) [inline projects/DiffusionDet/configs/diffusiondet_r50_fpn_500-proposals_crop-ms-480-800-450k_coco.py:97]:
```suggestion
        file_client_args=_base_.file_client_args,
```

File: projects/DiffusionDet/configs/diffusiondet_r50_fpn_500-proposals_crop-ms-480-800-450k_coco.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        imdecode_backend=backend),
+    dict(type='LoadAnnotations', with_bbox=True, with_mask=False),
```
Comment (hhaAndroid) [inline projects/DiffusionDet/configs/diffusiondet_r50_fpn_500-proposals_crop-ms-480-800-450k_coco.py:99]:
```suggestion
    dict(type='LoadAnnotations', with_bbox=True),
```

File: configs/_base_/datasets/coco_detection.py [hunk 6]
```diff
@@ -36,8 +36,8 @@
     dataset=dict(
         type=dataset_type,
         data_root=data_root,
-        ann_file='annotations/instances_train2017.json',
-        data_prefix=dict(img='train2017/'),
+        ann_file='annotations/instances_val2017.json',
```
Comment (RangiLyu) [inline configs/_base_/datasets/coco_detection.py:39]: delete this

File: projects/DiffusionDet/README.md (hunk 3 above)
Comment (BIGWangYuDong) [reply projects/DiffusionDet/README.md:62]: Seems This need keep to tell users, which step this project is done

File: projects/DiffusionDet/diffusiondet/head.py [hunk 7]
```diff
@@ -0,0 +1,957 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (RangiLyu) [inline projects/DiffusionDet/diffusiondet/head.py:1]: Need to add the origin repo's copyright.

File: projects/DiffusionDet/README.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+python projects/DiffusionDet/model_converters/diffusiondet_resnet_to_mmdet.py ${DiffusionDet ckpt path} ${MMDetectron ckpt path}
+```
+
+In MMDetection's root directory, run the following command to test the model:
+
+```bash
+python tools/test.py projects/DiffusionDet/model_converters/diffusiondet_resnet_to_mmdet.py ${CHECKPOINT_P
+```
+
+**Note:** During inference time, DiffusionDet will randomly generate noisy boxes, which may affect the AP results. If users want to get the same result every inference time, setting seed is a good way.
```
Comment (RangiLyu) [inline projects/DiffusionDet/README.md:41]: add the testing result chart

File: projects/DiffusionDet/README.md [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+## Usage
+
+<!-- For a typical model, this section should contain the commands for training and testing. You are also suggested to dump your environment specification to env.yml by `conda env export > env.yml`. -->
+
+### Training commands
+
+In MMDetection's root directory, run the following command to train the model:
+
+```bash
+python tools/train.py projects/DiffusionDet/model_converters/diffusiondet_resnet_to_mmdet.py
```
Comment (ZwwWayne) [inline projects/DiffusionDet/README.md:18]:
```suggestion
python projects/DiffusionDet/model_converters/diffusiondet_resnet_to_mmdet.py
```

File: projects/DiffusionDet/README.md [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+```
+
+For multi-gpu training, run:
+
+```bash
+python -m torch.distributed.launch --nnodes=1 --node_rank=0 --nproc_per_node=${NUM_GPUS} --master_port=29506 --master_addr="127.0.0.1" tools/train.py projects/DiffusionDet/model_converters/diffusiondet_resnet_to_mmdet.py
```
Comment (ZwwWayne) [inline projects/DiffusionDet/README.md:24]: better to provide test commands as we do not support training now

Comment (ZwwWayne) [reply projects/DiffusionDet/README.md:24]: or not provide it for now

File: projects/DiffusionDet/diffusiondet/diffusiondet.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+
+# ========================================
+# Modified by Shoufa Chen
+# ========================================
+# Modified by Peize Sun, Rufeng Zhang
+# Contact: {sunpeize, cxrfzhang}@foxmail.com
+#
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
```
Comment (ZwwWayne) [inline projects/DiffusionDet/diffusiondet/diffusiondet.py:10]: This file seems to be created by us, thus, we only need to keep the copyright of open-mmlab.

File: projects/DiffusionDet/README.md [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+1. Download the [DiffusionDet released model](https://github.com/ShoufaChen/DiffusionDet#models).
+
+2. Convert model from DiffusionDet version to MMDetection version. We give a [sample script](model_converters/diffusiondet_resnet_to_mmdet.py)
+   to convert `DiffusionDet-resnet50` model. Users can download the corresponding models from [here](https://github.com/ShoufaChen/DiffusionDet/releases/download/v0.1/diffdet_coco_res50.pth).
+
+   ```shell
+   python projects/DiffusionDet/model_converters/diffusiondet_resnet_to_mmdet.py ${DiffusionDet ckpt path} ${MMDetectron ckpt path}
+   ```
+
+3. Following \[Testing commands\](#Testing commands) to test the model in MMDetection.
```
Comment (BIGWangYuDong) [inline projects/DiffusionDet/README.md:24]:
```suggestion
3. Following [Testing commands](#Testing commands) to test the model in MMDetection.
```

Comment (BIGWangYuDong) [reply projects/DiffusionDet/README.md:24]: `mdformat` force adding `\` before `[ ]`, not sure the reason

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**